File: app/src/main/java/remix/myplayer/ui/adapter/SongAdapter.java
package remix.myplayer.ui.adapter;

import static remix.myplayer.request.ImageUriRequest.SMALL_IMAGE_SIZE;
import static remix.myplayer.theme.ThemeStore.getHighLightTextColor;
import static remix.myplayer.theme.ThemeStore.getTextColorPrimary;
import static remix.myplayer.util.ImageUriUtil.getSearchRequestWithAlbumType;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.PopupMenu;
import androidx.recyclerview.widget.RecyclerView;
import android.text.TextUtils;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import com.github.promeg.pinyinhelper.Pinyin;
import remix.myplayer.R;
import remix.myplayer.bean.mp3.Song;
import remix.myplayer.databinding.ItemSongRecycleBinding;
import remix.myplayer.databinding.LayoutHeader1Binding;
import remix.myplayer.helper.MusicServiceRemote;
import remix.myplayer.misc.menu.SongPopupListener;
import remix.myplayer.service.Command;
import remix.myplayer.theme.Theme;
import remix.myplayer.theme.ThemeStore;
import remix.myplayer.ui.adapter.holder.BaseViewHolder;
import remix.myplayer.ui.misc.MultipleChoice;
import remix.myplayer.ui.widget.fastcroll_recyclerview.FastScroller;
import remix.myplayer.util.MusicUtil;
import remix.myplayer.util.ToastUtil;

/**
 * 全部歌曲和最近添加页面所用adapter
 */

/**
 * Created by Remix on 2016/4/11.
 */
public class SongAdapter extends HeaderAdapter<Song, BaseViewHolder> implements
    FastScroller.SectionIndexer {

  private Song mLastPlaySong = MusicServiceRemote.getCurrentSong();

  public SongAdapter(int layoutId, MultipleChoice multiChoice, RecyclerView recyclerView) {
    super(layoutId, multiChoice, recyclerView);
    mRecyclerView = recyclerView;
  }

  @NonNull
  @Override
  public BaseViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    return viewType == TYPE_HEADER ?
        new HeaderHolder(
            LayoutInflater.from(parent.getContext())
                .inflate(R.layout.layout_header_1, parent, false)) :
        new SongViewHolder(
            LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_song_recycle, parent, false));
  }

  @Override
  public void onViewRecycled(@NonNull BaseViewHolder holder) {
    super.onViewRecycled(holder);
    disposeLoad(holder);
  }

  @SuppressLint("RestrictedApi")
  @Override
  protected void convert(BaseViewHolder baseHolder, final Song song, int position) {
    final Context context = baseHolder.itemView.getContext();
    if (position == 0) {
      final HeaderHolder headerHolder = (HeaderHolder) baseHolder;
      //没有歌曲时隐藏
      if (mDatas == null || mDatas.size() == 0) {
        headerHolder.binding.getRoot().setVisibility(View.GONE);
        return;
      } else {
        headerHolder.binding.getRoot().setVisibility(View.VISIBLE);
      }

      headerHolder.binding.playShuffleButton.setImageDrawable(
          Theme.tintVectorDrawable(context, R.drawable.ic_shuffle_white_24dp,
              ThemeStore.getAccentColor())
      );
      headerHolder.binding.tvShuffleCount.setText(context.getString(R.string.play_random, getItemCount() - 1));

      headerHolder.binding.getRoot().setOnClickListener(v -> {
        Intent intent = MusicUtil.makeCmdIntent(Command.NEXT, true);
        if (mDatas == null || mDatas.isEmpty()) {
          ToastUtil.show(context, R.string.no_song);
          return;
        }
        MusicServiceRemote.setPlayQueue(mDatas, intent);
      });
      return;
    }

    if (!(baseHolder instanceof SongViewHolder)) {
      return;
    }
    final SongViewHolder holder = (SongViewHolder) baseHolder;

    //封面
    holder.binding.songHeadImage.setTag(
        setImage(holder.binding.songHeadImage, getSearchRequestWithAlbumType(song), SMALL_IMAGE_SIZE, position));

//        //是否为无损
//        if(!TextUtils.isEmpty(song.getDisplayName())){
//            String prefix = song.getDisplayName().substring(song.getDisplayName().lastIndexOf(".") + 1);
//            holder.mSQ.setVisibility(prefix.equals("flac") || prefix.equals("ape") || prefix.equals("wav")? View.VISIBLE : View.GONE);
//        }

    //高亮
    if (MusicServiceRemote.getCurrentSong().getId() == song.getId()) {
      mLastPlaySong = song;
      holder.binding.songTitle.setTextColor(getHighLightTextColor());
      holder.binding.indicator.setVisibility(View.VISIBLE);
    } else {
      holder.binding.songTitle.setTextColor(getTextColorPrimary());
      holder.binding.indicator.setVisibility(View.GONE);
    }
    holder.binding.indicator.setBackgroundColor(getHighLightTextColor());

    //标题
    holder.binding.songTitle.setText(song.getShowName());

    //艺术家与专辑
    holder.binding.songOther.setText(String.format("%s-%s", song.getArtist(), song.getAlbum()));

    //设置按钮着色
    int tintColor = ThemeStore.getLibraryBtnColor();
    Theme.tintDrawable(holder.binding.songButton, R.drawable.icon_player_more, tintColor);

    holder.binding.songButton.setOnClickListener(v -> {
      if (mChoice.isActive()) {
        return;
      }
      final PopupMenu popupMenu = new PopupMenu(context, holder.binding.songButton, Gravity.END);
      popupMenu.getMenuInflater().inflate(R.menu.menu_song_item, popupMenu.getMenu());
      popupMenu.setOnMenuItemClickListener(
          new SongPopupListener((AppCompatActivity) context, song, false, ""));
      popupMenu.show();
    });

    holder.binding.itemRoot.setOnClickListener(v -> {
      if (position - 1 < 0) {
        ToastUtil.show(context, R.string.illegal_arg);
        return;
      }
      mOnItemClickListener.onItemClick(v, position - 1);
    });
    holder.binding.itemRoot.setOnLongClickListener(v -> {
      if (position - 1 < 0) {
        ToastUtil.show(context, R.string.illegal_arg);
        return true;
      }
      mOnItemClickListener.onItemLongClick(v, position - 1);
      return true;
    });

    holder.binding.itemRoot.setSelected(mChoice.isPositionCheck(position - 1));
  }

  @Override
  public String getSectionText(int position) {
    if (position == 0) {
      return "";
    }
    if (mDatas != null && position - 1 < mDatas.size()) {
      String title = mDatas.get(position - 1).getTitle();
      return !TextUtils.isEmpty(title) ? (Pinyin.toPinyin(title.charAt(0))).toUpperCase()
          .substring(0, 1) : "";
    }
    return "";
  }

  /**
   * 更新高亮歌曲
   */
  public void updatePlayingSong() {
    final Song currentSong = MusicServiceRemote.getCurrentSong();
    if (currentSong.getId() == -1 || currentSong.getId() == mLastPlaySong.getId()) {
      return;
    }

    if (mDatas != null && mDatas.contains(currentSong)) {
      // 找到新的高亮歌曲
      final int index = mDatas.indexOf(currentSong) + 1;
      final int lastIndex = mDatas.indexOf(mLastPlaySong) + 1;

      SongViewHolder newHolder = null;
      if (mRecyclerView.findViewHolderForAdapterPosition(index) instanceof SongViewHolder) {
        newHolder = (SongViewHolder) mRecyclerView.findViewHolderForAdapterPosition(index);
      }
      SongViewHolder oldHolder = null;
      if (mRecyclerView.findViewHolderForAdapterPosition(lastIndex) instanceof SongViewHolder) {
        oldHolder = (SongViewHolder) mRecyclerView.findViewHolderForAdapterPosition(lastIndex);
      }

      if (newHolder != null) {
        newHolder.binding.songTitle.setTextColor(getHighLightTextColor());
        newHolder.binding.indicator.setVisibility(View.VISIBLE);
      }

      if (oldHolder != null) {
        oldHolder.binding.songTitle.setTextColor(getTextColorPrimary());
        oldHolder.binding.indicator.setVisibility(View.GONE);
      }
      mLastPlaySong = currentSong;
    }
  }

  static class SongViewHolder extends BaseViewHolder {

    private final ItemSongRecycleBinding binding;

    SongViewHolder(View itemView) {
      super(itemView);
      binding = ItemSongRecycleBinding.bind(itemView);
    }
  }

  static class HeaderHolder extends BaseViewHolder {

    final LayoutHeader1Binding binding;

    HeaderHolder(View itemView) {
      super(itemView);
      binding = LayoutHeader1Binding.bind(itemView);
    }
  }
}

File: app/src/main/java/remix/myplayer/ui/adapter/SongAdapter.kt
package remix.myplayer.ui.adapter

import android.annotation.SuppressLint
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.PopupMenu
import androidx.recyclerview.widget.RecyclerView
import com.github.promeg.pinyinhelper.Pinyin
import remix.myplayer.App
import remix.myplayer.R
import remix.myplayer.bean.mp3.Song
import remix.myplayer.databinding.ItemSongRecycleBinding
import remix.myplayer.databinding.LayoutHeader1Binding
import remix.myplayer.helper.MusicServiceRemote.getCurrentSong
import remix.myplayer.helper.MusicServiceRemote.setPlayQueue
import remix.myplayer.helper.SortOrder
import remix.myplayer.misc.menu.SongPopupListener
import remix.myplayer.request.ImageUriRequest
import remix.myplayer.service.Command
import remix.myplayer.theme.Theme
import remix.myplayer.theme.ThemeStore.accentColor
import remix.myplayer.theme.ThemeStore.highLightTextColor
import remix.myplayer.theme.ThemeStore.libraryBtnColor
import remix.myplayer.theme.ThemeStore.textColorPrimary
import remix.myplayer.ui.adapter.holder.BaseViewHolder
import remix.myplayer.ui.misc.MultipleChoice
import remix.myplayer.ui.widget.fastcroll_recyclerview.FastScroller
import remix.myplayer.util.ImageUriUtil
import remix.myplayer.util.MusicUtil
import remix.myplayer.util.SPUtil
import remix.myplayer.util.ToastUtil
import java.util.*

/**
 * 全部歌曲和最近添加页面所用adapter
 */
/**
 * Created by Remix on 2016/4/11.
 */
class SongAdapter(layoutId: Int, multiChoice: MultipleChoice<Song>, recyclerView: RecyclerView)
  : HeaderAdapter<Song, BaseViewHolder>(layoutId, multiChoice, recyclerView), FastScroller.SectionIndexer {

  private var lastPlaySong = getCurrentSong()

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
    return if (viewType == TYPE_HEADER) HeaderHolder(LayoutInflater.from(parent.context).inflate(R.layout.layout_header_1, parent, false))
    else SongViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_song_recycle, parent, false))
  }

  override fun onViewRecycled(holder: BaseViewHolder) {
    super.onViewRecycled(holder)
    disposeLoad(holder)
  }

  @SuppressLint("RestrictedApi")
  override fun convert(holder: BaseViewHolder, song: Song?, position: Int) {
    val context = holder.itemView.context
    if (position == 0) {
      val headerHolder = holder as HeaderHolder
      //没有歌曲时隐藏
      if (dataList.size == 0) {
        headerHolder.binding.root.visibility = View.GONE
        return
      } else {
        headerHolder.binding.root.visibility = View.VISIBLE
      }
      headerHolder.binding.playShuffleButton.setImageDrawable(
          Theme.tintVectorDrawable(context, R.drawable.ic_shuffle_white_24dp,
              accentColor)
      )
      headerHolder.binding.tvShuffleCount.text = context.getString(R.string.play_random, itemCount - 1)
      headerHolder.binding.root.setOnClickListener { v: View? ->
        val intent = MusicUtil.makeCmdIntent(Command.NEXT, true)
        if (dataList.isEmpty()) {
          ToastUtil.show(context, R.string.no_song)
          return@setOnClickListener
        }
        setPlayQueue(dataList, intent)
      }
      return
    }
    if (holder !is SongViewHolder || song == null) {
      return
    }

    //封面
    holder.binding.songHeadImage.tag = setImage(holder.binding.songHeadImage, ImageUriUtil.getSearchRequestWithAlbumType(song), ImageUriRequest.SMALL_IMAGE_SIZE, position)

//        //是否为无损
//        if(!TextUtils.isEmpty(song.getDisplayName())){
//            String prefix = song.getDisplayName().substring(song.getDisplayName().lastIndexOf(".") + 1);
//            holder.mSQ.setVisibility(prefix.equals("flac") || prefix.equals("ape") || prefix.equals("wav")? View.VISIBLE : View.GONE);
//        }

    //高亮
    if (getCurrentSong().id == song.id) {
      lastPlaySong = song
      holder.binding.songTitle.setTextColor(highLightTextColor)
      holder.binding.indicator.visibility = View.VISIBLE
    } else {
      holder.binding.songTitle.setTextColor(textColorPrimary)
      holder.binding.indicator.visibility = View.GONE
    }
    holder.binding.indicator.setBackgroundColor(highLightTextColor)

    //标题
    holder.binding.songTitle.text = song.showName

    //艺术家与专辑
    holder.binding.songOther.text = String.format("%s-%s", song.artist, song.album)

    //设置按钮着色
    val tintColor = libraryBtnColor
    Theme.tintDrawable(holder.binding.songButton, R.drawable.icon_player_more, tintColor)
    holder.binding.songButton.setOnClickListener { v: View? ->
      if (choice.isActive) {
        return@setOnClickListener
      }
      val popupMenu = PopupMenu(context, holder.binding.songButton, Gravity.END)
      popupMenu.menuInflater.inflate(R.menu.menu_song_item, popupMenu.menu)
      popupMenu.setOnMenuItemClickListener(
          SongPopupListener((context as AppCompatActivity), song, false, ""))
      popupMenu.show()
    }
    holder.binding.itemRoot.setOnClickListener { v: View? ->
      if (position - 1 < 0) {
        ToastUtil.show(context, R.string.illegal_arg)
        return@setOnClickListener
      }
      onItemClickListener?.onItemClick(v, position - 1)
    }
    holder.binding.itemRoot.setOnLongClickListener { v: View? ->
      if (position - 1 < 0) {
        ToastUtil.show(context, R.string.illegal_arg)
        return@setOnLongClickListener true
      }
      onItemClickListener?.onItemLongClick(v, position - 1)
      true
    }
    holder.binding.itemRoot.isSelected = choice.isPositionCheck(position - 1)
  }

  override fun getSectionText(position: Int): String {
    if (position in 1..dataList.size) {
      val data = dataList[position - 1]
      val key = when (SPUtil.getValue(
        App.getContext(),
        SPUtil.SETTING_KEY.NAME,
        SPUtil.SETTING_KEY.SONG_SORT_ORDER,
        SortOrder.SONG_A_Z
      )) {
        SortOrder.SONG_A_Z, SortOrder.SONG_Z_A -> data.title
        SortOrder.ARTIST_A_Z, SortOrder.ARTIST_Z_A -> data.artist
        SortOrder.ALBUM_A_Z, SortOrder.ALBUM_Z_A -> data.album
        SortOrder.DISPLAY_NAME_A_Z, SortOrder.DISPLAY_NAME_Z_A -> data.displayName
        else -> ""
      }
      if (key.isNotEmpty())
        return Pinyin.toPinyin(key[0]).toUpperCase(Locale.getDefault()).substring(0, 1)
    }
    return ""
  }

  /**
   * 更新高亮歌曲
   */
  fun updatePlayingSong() {
    val currentSong = getCurrentSong()
    if (currentSong.id == -1 || currentSong.id == lastPlaySong.id) {
      return
    }
    if (dataList.contains(currentSong)) {
      // 找到新的高亮歌曲
      val index = dataList.indexOf(currentSong) + 1
      val lastIndex = dataList.indexOf(lastPlaySong) + 1
      var newHolder: SongViewHolder? = null
      if (recyclerView.findViewHolderForAdapterPosition(index) is SongViewHolder) {
        newHolder = recyclerView.findViewHolderForAdapterPosition(index) as SongViewHolder?
      }
      var oldHolder: SongViewHolder? = null
      if (recyclerView.findViewHolderForAdapterPosition(lastIndex) is SongViewHolder) {
        oldHolder = recyclerView.findViewHolderForAdapterPosition(lastIndex) as SongViewHolder?
      }
      if (newHolder != null) {
        newHolder.binding.songTitle.setTextColor(highLightTextColor)
        newHolder.binding.indicator.visibility = View.VISIBLE
      }
      if (oldHolder != null) {
        oldHolder.binding.songTitle.setTextColor(textColorPrimary)
        oldHolder.binding.indicator.visibility = View.GONE
      }
      lastPlaySong = currentSong
    }
  }

  internal class SongViewHolder(itemView: View) : BaseViewHolder(itemView) {
    val binding: ItemSongRecycleBinding = ItemSongRecycleBinding.bind(itemView)
  }

  internal class HeaderHolder(itemView: View) : BaseViewHolder(itemView) {
    val binding: LayoutHeader1Binding = LayoutHeader1Binding.bind(itemView)

  }

}